### PR TITLE
Fix COPY; rework Reader op

### PR DESCRIPTION
**Pull request by ray6080**

### CODE REVIEW

File: src/include/storage/copier/rel_copier.h [hunk 1]
```diff
@@ -1,6 +1,8 @@
 #pragma once
 
-#include "storage/copier/read_file_state.h"
+#include <utility>
```
Comment (ray6080) [inline src/include/storage/copier/rel_copier.h:3]: Remove this include.

File: src/storage/copier/reader_state.cpp [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+std::vector<FileBlocksInfo> ReaderFunctions::countRowsInParquetFile(
+    std::vector<std::string>& paths, CSVReaderConfig csvReaderConfig, TableSchema* tableSchema) {
+    std::vector<FileBlocksInfo> result(paths.size());
+    for (auto i = 0u; i < paths.size(); i++) {
+        std::unique_ptr<parquet::arrow::FileReader> reader =
+            TableCopyUtils::createParquetReader(paths[i], tableSchema);
+        auto metadata = reader->parquet_reader()->metadata();
+        auto numBlocks = metadata->num_row_groups();
+        std::vector<row_idx_t> blocks;
+        blocks.resize(numBlocks);
```
Comment (acquamarin) [inline src/storage/copier/reader_state.cpp:106]: use reserve and call push_back() in the for loop below?

File: src/storage/copier/reader_state.cpp [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    case (CopyDescription::FileType::NPY):
+        return validateNPYFiles;
+    default:
+        throw NotImplementedException{"ReaderFunctions::getValidateFunc"};
+    }
+}
+
+count_blocks_func_t ReaderFunctions::getCountBlocksFunc(CopyDescription::FileType fileType) {
+    switch (fileType) {
+    case (CopyDescription::FileType::CSV):
```
Comment (acquamarin) [inline src/storage/copier/reader_state.cpp:24]: ditto, remove bracket

File: src/storage/copier/reader_state.cpp [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+using namespace kuzu::catalog;
+using namespace kuzu::common;
+
+namespace kuzu {
+namespace storage {
+
+validate_func_t ReaderFunctions::getValidateFunc(CopyDescription::FileType fileType) {
+    switch (fileType) {
+    case (CopyDescription::FileType::CSV):
```
Comment (acquamarin) [inline src/storage/copier/reader_state.cpp:11]: I think our convention is not to add a bracket to the case statement.

**Review by acquamarin — APPROVED** [review]

File: src/processor/operator/persistent/reader.cpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+bool Reader::getNextTuplesInternal(ExecutionContext* context) {
+    std::shared_ptr<arrow::Table> table = nullptr;
+    readerInfo.isOrderPreserving ? getNextNodeGroupInSerial(table) :
+                                   getNextNodeGroupInParallel(table);
+    if (table == nullptr) {
+        return false;
+    }
+    for (auto i = 0u; i < readerInfo.dataColumnPoses.size(); i++) {
+        ArrowColumnVector::setArrowColumn(
```
[the remark this answers is not in the document]
Comment (andyfengHKU) [reply src/processor/operator/persistent/reader.cpp:20]: I'm changing this as well. Let me do this.

File: src/include/storage/copier/column_chunk.h [hunk 6]
```diff
@@ -16,31 +16,34 @@ namespace storage {
...
 class NullColumnChunk;
 
-struct ColumnChunkMetadata {
-    common::page_idx_t pageIdx = common::INVALID_PAGE_IDX;
-    common::page_idx_t numPages = 0;
-
-    ColumnChunkMetadata() = default;
-    ColumnChunkMetadata(common::page_idx_t pageIdx, common::page_idx_t numPages)
+struct BaseColumnChunkMetadata {
+    BaseColumnChunkMetadata() : pageIdx{common::INVALID_PAGE_IDX}, numPages{0} {}
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/include/storage/copier/column_chunk.h:20]: hmm ok. Will make the change.
Was this style coming from some standard? I wasn't aware of that. why do we have to make struct so different from class?

File: src/processor/operator/persistent/reader.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ray6080) [reply src/processor/operator/persistent/reader.cpp:20]: OK. I will leave this to @andyfengHKU .
But i think it's better that we just store either column pos or value vector, as these two are redundant. I would prefer storing column pos, and dynamically fetch the value vector. We can discuss a bit more.

**Pull request merged**